### PR TITLE
[FIX] sale_margin: Correct computation of margin in groupby

### PR DESCRIPTION
**Pull request by abz-odoo**
For now the sale margin is incorrectly computed in group by

opw-2767397

### CODE REVIEW

**Review by Feyensv — APPROVED** [review]
Seems like the demo data of sale should be improved one day :sweat_smile: 
image

robodoo r+

File: addons/sale_margin/models/sale_order.py [hunk 1]
```diff
@@ -10,7 +10,7 @@ class SaleOrderLine(models.Model):
         "Margin", compute='_compute_margin',
         digits='Product Price', store=True, groups="base.group_user")
     margin_percent = fields.Float(
-        "Margin (%)", compute='_compute_margin', store=True, groups="base.group_user")
+        "Margin (%)", compute='_compute_margin', store=True, groups="base.group_user", group_operator="avg")
```
Comment (fmdl) [inline addons/sale_margin/models/sale_order.py:13]: the avg operator is not good.
you should by pass the readgroup function.

exemple:
line 1 : cost 10, price 20 : marge 50%
line 2: cost 10, price 100 : marge 90%

avg(50, 90) --> 70%
(20 + 100 - 10 - 10)/(20 + 100) --> 83%

The good value is 83%

@tde-banana-odoo

**Review by Feyensv — CHANGES_REQUESTED** [review]
Indeed, we should hide the field in the graph view instead, my bad.
(See https://github.com/odoo/odoo/pull/60290, same issue, but solved only for the pivot view, the graph view was forgotten ^^)

File: addons/sale_margin/models/sale_order.py (hunk 1 above)
Comment (bouvyd) [reply addons/sale_margin/models/sale_order.py:13]: the point is not to have an exact computation in all cases, the point is to have at least one computation that makes sense
e.g. you can say "on average, for customer X my margin is X%" (when grouping by partner, for ex.)
summing will *never* be correct though

overriding the read_group seems like overkill for this (imho)

although perhaps @dala should have the last word on this

Comment (Feyensv) [reply addons/sale_margin/models/sale_order.py:13]: it's @dalaOdoo I think :eyes:

**Comment by fmdl** [comment]
please find the solution (should same in sale.report and sale.order.line)

```python
    @api.model
    def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True):
        if 'margin_percent' in fields:
              fields = list(set( fields + ['amount_untaxed', 'margin']))
        res = super().read_group(domain, fields, groupby, offset=offset, limit=limit, orderby=orderby, lazy=lazy)
        for r in res:
            if r['amount_untaxed']:
                r['margin_percent'] = 100 * r['margin'] / r['amount_untaxed']
            else:
                r.pop('margin_percent', None)
        return res
```

File: addons/sale_margin/models/sale_order.py (hunk 1 above)
Comment (dalaOdoo) [reply addons/sale_margin/models/sale_order.py:13]: The margin % is not available as a measure in the sale reporting and in the pivot view so there is no reason to allow it in the graph view at this stage. The easiest fix is to remove it from the graph view.

**Review by Feyensv — APPROVED** [review]
@robodoo r+

File: addons/sale_margin/models/sale_order.py (hunk 1 above)
Comment (fmdl) [inline addons/sale_margin/models/sale_order.py:13]: can you remove the avg ?

Comment (Feyensv) [reply addons/sale_margin/models/sale_order.py:13]: It's kept for the specific case where users add the `margin_percent` in a SO list view.

Comment (fmdl) [reply addons/sale_margin/models/sale_order.py:13]: but if you group, the value is wrong (it is better to show 13400% than a 70% wrong value)

Comment (fmdl) [reply addons/sale_margin/models/sale_order.py:13]: what do you think about my fix ?

```python
    @api.model
    def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True):
        if 'margin_percent' in fields:
              fields = list(set( fields + ['amount_untaxed', 'margin']))
        res = super().read_group(domain, fields, groupby, offset=offset, limit=limit, orderby=orderby, lazy=lazy)
        for r in res:
            if r['amount_untaxed']:
                r['margin_percent'] = 100 * r['margin'] / r['amount_untaxed']
            else:
                r.pop('margin_percent', None)
        return res
```

Comment (Feyensv) [reply addons/sale_margin/models/sale_order.py:13]: > the point is not to have an exact computation in all cases, the point is to have at least one computation that makes sense
> e.g. you can say "on average, for customer X my margin is X%" (when grouping by partner, for ex.)
> 
> overriding the read_group seems like overkill for this (imho)

Comment (pedrobaeza) [reply addons/sale_margin/models/sale_order.py:13]: @bouvyd are you around again?!

This is the proper way to compute such margin, that we are doing on invoices:

https://github.com/OCA/margin-analysis/blob/13.0/account_invoice_margin/report/account_invoice_report.py

Comment (Feyensv) [reply addons/sale_margin/models/sale_order.py:13]: He's back as Product Owner now (not a dev anymore) :rocket:

**Pull request closed without merging**